hardware_interface/test/test_handle.cpp

```diff
@@ -68,3 +75,33 @@ TEST(TestHandle, value_methods_work_on_non_nullptr)
   EXPECT_NO_THROW(handle.set_value(0.0));
   EXPECT_DOUBLE_EQ(handle.get_value(), 0.0);
 }
+
+TEST(TestHandle, no_data_race_when_accessing_value) // fails when used with regular handles due to thread sanitizer
+{
+  std::atomic<double> state_if_value = 1.558;
+  std::atomic<double> cmd_if_value = 1.337;
+
+  AsyncStateInterface state_handle{JOINT_NAME, FOO_INTERFACE, &state_if_value};
+  AsyncCommandInterface command_handle{JOINT_NAME, FOO_INTERFACE, &cmd_if_value};
```

Comment:
As far as I know it's not possible for two controllers to access the same cmd interfaces at once. But if you meant a sync hardware component, your point is valid, it's not guarranted that the controller will see the latest values. In fact, something like this could easily happen in case of using an asynchronous JTC on a 6-dof robot arm, where each joint is represented by a single command interface:

- Joint1: async JTC sees the most recent value
- Joint2: async JTC sees the most recent value
- Joint3: async JTC sees the most recent value
- Joint4: async JTC sees the most recent value
- Joint5: async JTC sees the previous value
- Joint6: async JTC sees the previous value

And the trajectory could get executed with joint5 and joint6 containing values from a previous tick. I guess it depends on the use case if this is acceptable or not. 
I was thinking about a solution which solves this by notifying the async controller when the read is done (just as you've suggested previously), BUT the writes from the async controller to the command interfaces would still be non-blocking relaxed atomics. This would ensure that the controller works with the latest data, but the control loop is never blocked by the asynchronous controller if it takes forever to write its interfaces.

Would something like this fit your use case?

The AsyncCommandInterface in this example sadly can't assure atomicity of all used command interfaces, since it only represents a single command interface. I might have misunderstood you here though.